ledger/streaming/client/src/backfill.rs

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! A BlockStream that backfills missing indices from a given BlockFetcher.
+
+use displaydoc::Display;
+use futures::{FutureExt, Stream, StreamExt};
+use mc_common::logger::{log, Logger};
+use mc_ledger_streaming_api::{BlockFetcher, BlockStream, BlockStreamComponents, Result};
+use mc_transaction_core::BlockIndex;
+use std::{pin::Pin, sync::Arc};
+
+/// A [BlockStream] that backfills another [BlockStream] using a [BlockFetcher].
+#[derive(Debug, Display)]
+pub struct BackfillingStream<S: BlockStream + 'static, F: BlockFetcher + 'static> {
+    upstream: S,
+    fetcher: Arc<F>,
+    logger: Logger,
+}
+
+impl<S: BlockStream + 'static, F: BlockFetcher + 'static> BackfillingStream<S, F> {
+    /// Instantiate a [BackfillingStream].
+    pub fn new(upstream: S, fetcher: F, logger: Logger) -> Self {
+        Self {
+            upstream,
+            fetcher: Arc::new(fetcher),
+            logger,
+        }
+    }
+}
+
+impl<S: BlockStream + 'static, F: BlockFetcher + 'static> BlockStream for BackfillingStream<S, F> {
+    type Stream = impl Stream<Item = Result<BlockStreamComponents>>;
+
+    fn get_block_stream(&self, starting_height: u64) -> Result<Self::Stream> {
+        self.upstream
+            .get_block_stream(starting_height)
+            .map(|upstream| {
+                backfill_stream(
+                    upstream,
+                    starting_height,
+                    self.fetcher.clone(),
+                    self.logger.clone(),
+                )
+            })
+    }
+}
+
+fn backfill_stream<
+    S: Stream<Item = Result<BlockStreamComponents>> + 'static,
+    F: BlockFetcher + 'static,
+>(
+    upstream: S,
+    starting_height: u64,
+    fetcher: Arc<F>,
+    logger: Logger,
+) -> impl Stream<Item = Result<BlockStreamComponents>> {
+    use futures::stream::{empty, once};
+
+    let mut prev_index: Option<BlockIndex> = None;
+    upstream.flat_map(
+        move |result| -> Pin<Box<dyn Stream<Item = Result<BlockStreamComponents>>>> {
+            let next_index = prev_index.map_or_else(|| starting_height, |index| index + 1);
+            match result {
+                Ok(components) => {
+                    let index = components.block_data.block().index;
+                    if prev_index.is_some() && index <= prev_index.unwrap() {
+                        log::info!(
+                            logger,
+                            "Ignoring streamed block with index={} since last fetched index={}",
+                            index,
+                            prev_index.unwrap()
+                        );
+                        return Box::pin(empty());
+                    }
+
+                    let item_stream = once(async { Ok(components) });
+                    if index == next_index {
+                        prev_index = Some(index);
+                        Box::pin(item_stream)
```

Comment:
We could potentially use Pin-project here as it is meant to generating pinning in optimized ways

Reply:
I don't think so, since the Box here is to use different types of streams (via `dyn`), not operating on members of a struct.